feat(curriculum): add factorial lab (#55889)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/lab-factorial-calculator/66c07238b01053abaf812065.md

---
id: 66c07238b01053abaf812065
title: Build a Factorial Calculator
challengeType: 14
dashedName: build-a-factorial-calculator
---

# --description--

> **Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. 

Reviewer: We should add a sentence or two explaining what a factorial is because a lot of campers might not know or remember from math class years ago

**User Stories:**

> 1. You should ask the user for a number using a prompt window and store the number in a variable called `inputNumber`.

Reviewer: I think this should be simplified to ask for the prompt method instead of prompt window

```suggestion
1. You should ask the user for a number using the `prompt` method and store the number in a variable called `inputNumber`.
```

> 1. You should ask the user for a number using a prompt window and store the number in a variable called `inputNumber`.

Reviewer: We should use method here instead of prompt window

```suggestion
1. You should ask the user for a number using the `prompt` method and store the number in a variable called `inputNumber`.
```

   
2. Declare a variable `result` and assign it the value of `1`.
   

> 3. Using a for loop, you should loop through all numbers from `1` to `inputNumber` (inclusive) and for each number, multiply the `result` variable by the current number and assign the result to the `result` variable.

Reviewer: we should allow for any type of loop here.
we should only care that they use some sort of loop and get the correct final result

```suggestion
3. Using a loop, loop through all numbers from `1` to `inputNumber` (inclusive) and for each number, multiply the `result` variable by the current number and assign the result to the `result` variable. You can choose to use  either a`for` loop, `while` loop or `do...while` loop here.
```

   
4. You should log the result of the factorial of `inputNumber` to the console. The output should be in the format `The factorial of [inputNumber] is [result]`.

# --hints--

Your `link` element should be within your `head` element.

```js

```

# --seed--

## --seed-contents--

```js

```

# --solutions--

```js
let inputNumber = prompt("Enter a number: ");
let result = 1;
for (let i = 1; i <= inputNumber; i++) {
  result *= i;
}

> console.log(`Factorial of ${inputNumber} is: ${result}`); 

Reviewer: I was playing around with this locally and I noticed a few things missing.

Technically the user can input anything

<img width="950" alt="Screenshot 2024-08-19 at 11 21 48 AM" src="https://github.com/user-attachments/assets/22532e62-7e71-401f-b43a-7ecd105ad6c5">

But also, prompt will return a string
<img width="952" alt="Screenshot 2024-08-19 at 11 22 44 AM" src="https://github.com/user-attachments/assets/010bf3c7-9af4-4d9f-85e5-4c4f7b22347e">

So this solution as well as the user stories will need to be updated to check for the following:

- we should have campers write an condition to check that the input is not an empty string
- then campers will need to convert the input to a number and check that it is not NaN

<img width="844" alt="Screenshot 2024-08-19 at 11 26 30 AM" src="https://github.com/user-attachments/assets/00dfb2d5-c22b-4e61-9653-8c44b71c30c9">

if the user does not provide a valid input for either of those conditions above, then they should be prompted again until they do provide a valid input.

Reviewer: campers will also know about isNaN at that point if you wanted them to check using that

**User Stories:**

A factorial is the product of an integer and all the integers below it. For example, the factorial of `5` is `5 * 4 * 3 * 2 * 1 = 120`. In this lab, you will create a factorial calculator that takes a number from the user and calculates the factorial of that number.

1. You should declare a variable `num` and assign it the value of `5`.
   
2. Declare a variable `result` and assign it the value of `1`.
   
3. Using a loop, loop through all numbers from `1` to `num` (inclusive) and for each number, multiply the `result` variable by the current number and assign the result to the `result` variable. You can choose to use  either a `for` loop, `while` loop or `do...while` loop here.
   
4. You should store the factorial of `num` in the variable `resultMsg` in the format `Factorial of [num] is [result]`.

5. You should output the value of `resultMsg` to the console.
   
# --hints--

You should have a `num` variable.

```js
assert.isDefined(num);
```

You should assign `5` to the `num` variable

```js
assert.equal(num, 5);
```

You should have a `result` variable.

```js
assert.isDefined(result);
```

You should assign `1` to the `result` variable.

```js

> assert.match(__helpers.removeJSComments(code), /result\s*=\s*1\s*;?\s?$/m);

Reviewer: do we need a regex here?
assert.equal should work since result is a global variable right?

Author: `assert.equal` didn't work here as expected, maybe because it is reassigned later on.

---
id: 66c07238b01053abaf812065
title: Build a Factorial Calculator
challengeType: 14
dashedName: build-a-factorial-calculator
---

# --description--

**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab.

**User Stories:**

A factorial is the product of an integer and all the integers below it. For example, the factorial of `5` is `5 * 4 * 3 * 2 * 1 = 120`. In this lab, you will create a factorial calculator that takes a number from the user and calculates the factorial of that number.

> 1. You should declare a variable `num` and assign it the value of `5`.

Reviewer 2: I would ask to assign a generic number here

Author: After discussion in the chat room, we decided to use a predefined number here.

Reviewer: One of the reasons why I suggested using 5 is because that is a reasonable number for a factorial.

What we don't want is for the user to pick 1,000,000 or something like that.

If we want users to pick their own number we do need to put limits. So I vote the limit be 20. That seems reason for this exercise.

Then we would just need to test for the following:
- `num` should actually be a number and not some sort of string, or anything else
- the number needs to be a positive number between 1 -20

those should be easy to test with built in chai assert methods

   
2. Declare a variable `result` and assign it the value of `1`.
   

> 3. Using a loop, loop through all numbers from `1` to `inputNumber` (inclusive) and for each number, multiply the `result` variable by the current number and assign the result to the `result` variable. You can choose to use  either a `for` loop, `while` loop or `do...while` loop here.

Reviewer 2: ```suggestion
3. Using a loop, loop through all numbers from `1` to `num` (inclusive) and for each number, multiply the `result` variable by the current number and assign the result to the `result` variable. You can choose to use  either a `for` loop, `while` loop or `do...while` loop here.
```

   
4. You should store the factorial of `num` in the variable `resultMsg` in the format `Factorial of [num] is [result]`.

5. You should output the value of `resultMsg` to the console.
   
# --hints--

You should have a`num` variable

```js
assert.isDefined(num);
```

You should assign `5` to the `num` variable

```js
assert.equal(num, 5);
```

You should have a `result` variable.

```js
assert.isDefined(result);
```

You should assign `1` to the `result` variable.

```js
assert.match(__helpers.removeJSComments(code), /result\s*=\s*1;/);
```

The factorial of `5` should be `120`.

```js
const num = 5

let resultExpected = 1;

for (let i = 1; i <= num; i++) {
  resultExpected *= i;
}
assert.equal(result, resultExpected);

```

You should assign the final result to the variable `resultMsg` in the format `Factorial of [num] is [result]`.

```js

> assert.equal(resultMsg, 'Factorial of 5 is: 120');

Reviewer 2: ```suggestion
assert.equal(resultMsg, 'Factorial of 5 is 120');
```
